owning-a-smart-contract.md

```diff
@@ -0,0 +1,423 @@
+  gas: 200000,
+  ttl: 55
+}
+
+describe('Restricted', () => {
+
+  let owner;
+  let nonOwner;
+  let contractSource;
+
+  before(async () => {
+
+    // Create client objects
+    owner = await Ae({
+      url: config.host,
+      internalUrl: config.internalHost,
+      keypair: wallets[0],
+      nativeMode: true,
+      networkId: 'ae_devnet'
+    });
+
+    nonOwner = await Ae({
+      url: config.host,
+      internalUrl: config.internalHost,
+      keypair: wallets[1],
+      nativeMode: true,
+      networkId: 'ae_devnet'
+    });
+
+    // Read the source file
+    contractSource = utils.readFileRelative(config.contractSourceFile, "utf-8");
+
+  })
+
+  it('deploying successfully', async () => {
+    // Compile the contract
+    const compiledContract = await owner.contractCompile(contractSource, {
+      gas: config.gas
+    })
+
+    // Deploy the contract
+    const deployPromise = compiledContract.deploy({
+      options: {
+        ttl: config.ttl,
+      },
+      abi: "sophia"
+    });
+    // See if it has deployed correctly
+    assert.isFulfilled(deployPromise, 'Could not deploy the Restricted Smart Contract');
+  })
+
+  describe('Calling functions', () => {
+    let compiledContract;
+    let deployedContract;
+
+    before(async () => {
+      compiledContract = await owner.contractCompile(contractSource, {
+        gas: config.gas
+      })
+
+      deployedContract = await compiledContract.deploy({
+        options: {
+          ttl: config.ttl
+        },
+        abi: "sophia"
+      });
+    })
+
+    it('Should successfully call the non restricted method', async () => {
+      // Call the non restricted method
+      const callNotRestrictedMethod = nonOwner.contractCall(compiledContract.bytecode, 'sophia', deployedContract.address, "nonRestrictedFunction", {
+        args: `(2)`,
+        options: {
+          ttl: config.ttl
+        },
+        abi: "sophia"
+      })
+
+      await assert.isFulfilled(callNotRestrictedMethod, 'Calling the non restricted function failed');
+      const callResult = await callNotRestrictedMethod;
+      const decodedData = await callResult.decode('int') // get and decode the result
+
+      assert.equal(decodedData.value, 2, 'The returned data was not correct')
+    })
+
+    it('Should successfully call the restricted method', async () => {
+      // Call the restricted method from non owner
+      const callRestrictedMethod = nonOwner.contractCall(compiledContract.bytecode, 'sophia', deployedContract.address, "restrictedFunction", {
+        args: `(2)`,
+        options: {
+          ttl: config.ttl
+        },
+        abi: "sophia"
+      })
+
+      // It should be rejected as it is restricted to owner only
+      await assert.isRejected(callRestrictedMethod);
+    })
+  })
+})
+```
+
+As we have not yet implemented the restriction functionality the third test naturally fails.
+Run the following command in the terminal:
+```
+aeproject test
+```
+
+```cmake=
```

Comment:
when I replace the `exampleTest.js` with it  the tests fail:

```
===== Starting Tests =====


  Restricted
    1) "before all" hook


  0 passing (3ms)
  1 failing

  1) Restricted
       "before all" hook:
     ReferenceError: wallets is not defined
      at Context.before (test/exampleTest.js:23:16)
      at process.topLevelDomainCallback (domain.js:120:23)
```

Reply:
aeproject PR is now merged